### PR TITLE
NFT expanded states v2

### PR DESCRIPTION
**Pull request by christianbaroni**
todo:
- [x] cache NFT image metadata: `aspectRatio`, `imageColor`
- [x] fix VibrancyView render artifacts
- [x] Android action sheets
- [x] test Android

nice to have:
- [ ] get imageColor from SVG
- [ ] find a way to remove the SVG loading flash
- [ ] freeze blurred GIFs in expanded state backgrounds
- [ ] cache floor prices and share across collections
- [ ] don't check a collection's floor price on sheet open if it's been checked within last ~30 secs

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -77,6 +77,7 @@
     "@rainbow-me/react-native-payments": "1.1.5",
     "@react-native-community/async-storage": "1.12.1",
     "@react-native-community/blur": "3.6.0",
+    "@react-native-community/cameraroll": "^4.1.2",
```
Comment (jinchung) [inline package.json:80]: hey @brunobar79 do we have a pattern now of locking versions we want to apply for all new dependencies

File: src/components/sheet/SlackSheet.js [hunk 2]
```diff
@@ -92,11 +92,13 @@ export default forwardRef(function SlackSheet(
...
     testID,
     removeClippedSubviews = false,
+    yPosition: givenYPosition,
     ...props
   },
   ref
 ) {
-  const yPosition = useSharedValue(0);
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  const yPosition = givenYPosition || useSharedValue(0);
```
Comment (jinchung) [inline src/components/sheet/SlackSheet.js:101]: oh helllllllll no :)

File: src/components/asset-list/RecyclerAssetList/index.tsx [hunk 3]
```diff
@@ -769,66 +766,6 @@ function RecyclerAssetList({
       const colleciblesStartHeight =
         balancesHeight + smallBalancesHeight + savingsHeight + poolsHeight;
 
-      // Auto-scroll to opened family logic 👇
```
Comment (jinchung) [inline src/components/asset-list/RecyclerAssetList/index.tsx:772]: Where's this going? @osdnk @skylarbarrera

File: src/components/sheet/SlackSheet.js (hunk 2 above)
Comment (christianbaroni) [reply src/components/sheet/SlackSheet.js:101]: @osdnk

File: src/components/expanded-state/unique-token/ZoomableWrapper.js [hunk 4]
```diff
@@ -0,0 +1,623 @@
...
+  isENS,
+  borderRadius,
+  disableAnimations,
+  yDisplacement: givenYDisplacement,
+}) => {
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  const animationProgress = givenAnimationProgress || useSharedValue(0);
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  const yDisplacement = givenYDisplacement || useSharedValue(0);
+
```
Comment (brunobar79) [inline src/components/expanded-state/unique-token/ZoomableWrapper.js:80]: I can see why are you doing this, I wonder if there's an alternative.

File: src/components/expanded-state/unique-token/saveToCameraRoll.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import CameraRoll from '@react-native-community/cameraroll';
+import { Alert, PermissionsAndroid, Platform } from 'react-native';
+import RNFetchBlob from 'rn-fetch-blob';
+
+const getPermissionAndroid = async () => {
+  try {
+    const granted = await PermissionsAndroid.request(
```
Comment (brunobar79) [inline src/components/expanded-state/unique-token/saveToCameraRoll.js:7]: @ibrahimtaveras00  make sure you test this on both android & iOS (Save to camera roll)

File: src/components/sheet/SlackSheet.js (hunk 2 above)
Comment (brunobar79) [reply src/components/sheet/SlackSheet.js:101]: actually I understand the reasoning so we might have no other choice

File: src/hooks/usePersistentDominantColorFromImage.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  id,
+});
+
+type Result = {
+  state: State;
+  result: string;
+};
+
+const size = Math.ceil(CardSize) * PixelRatio.get();
+const getLowResUrl = (url: string) => {
```
Comment (brunobar79) [inline src/hooks/usePersistentDominantColorFromImage.ts:26]: This function is duplicated here: https://github.com/rainbow-me/rainbow/pull/2472/files#diff-ec86c9777ca223963133e263abd68388985faf046411af00fbf77ad2cc264aebR27 so it should be moved to a common place and reused

File: src/components/unique-token/UniqueTokenCard.js [hunk 7]
```diff
@@ -35,20 +44,30 @@ const UniqueTokenCard = ({
   width,
   ...props
 }) => {
+  const size = Math.ceil(CardSize) * pixelRatio;
+
+  const lowResUrl = useMemo(() => {
```
Comment (brunobar79) [inline src/components/unique-token/UniqueTokenCard.js:49]: This is also here https://github.com/rainbow-me/rainbow/pull/2472/files#diff-ec86c9777ca223963133e263abd68388985faf046411af00fbf77ad2cc264aebR27

We should extract it and reuse everywhere

File: src/parsers/uniqueTokens.js [hunk 8]
```diff
@@ -55,12 +65,20 @@ export const parseAccountUniqueTokens = data => {
         asset_contract.nft_version === '1.0' ||
         asset_contract.nft_version === '3.0' ||
         asset_contract.schema_name === 'ERC1155',
-      lastPrice: asset.last_sale ? Number(asset.last_sale.total_price) : null,
+      lastPrice: asset.last_sale
+        ? Number(asset.last_sale.total_price / 1000000000000000000) +
+          ` ${asset.last_sale.payment_token.symbol}`
```
Comment (brunobar79) [inline src/parsers/uniqueTokens.js:70]: use optional here

File: src/parsers/uniqueTokens.js [hunk 9]
```diff
@@ -55,12 +65,20 @@ export const parseAccountUniqueTokens = data => {
...
+        : null,
+      lastPriceUsd: asset.last_sale
+        ? asset.last_sale.payment_token.usd_price
```
Comment (brunobar79) [inline src/parsers/uniqueTokens.js:73]: optional just to be safe

File: src/react-native-cool-modals/ios/UIViewController+slack.swift [hunk 10]
```diff
@@ -124,7 +122,10 @@ class PanModalViewController: UIViewController, PanModalPresentable, UILayoutSup
   }
 
   var cornerRadius: CGFloat {
-    return CGFloat(truncating: config?.cornerRadius ?? 0)
+    if (self.config?.cornerRadius == 0.666) {
```
Comment (brunobar79) [inline src/react-native-cool-modals/ios/UIViewController+slack.swift:125]: 😈

**Review by brunobar79 — APPROVED** [review]
Added some minor comments but  💯

**Review by ibrahimtaveras00 — DISMISSED** [review]
**Issue 1:** 

NFT image is not appearing on wallet view once expanded only on android 

android (left) iOS (right)
Screen Shot 2021-11-04 at 10 53 44 AM

**Issue 2:**

None of the menu taps on Android are updated, shows old modal (copy contract address, view on etherscan and cancel) 

android (left) iOS (right)
Screen Shot 2021-11-04 at 10 55 39 AM

**Issue 3:**

Android only; tapping Share and tapping copy doesn't actually copy to clipboard 

image

**Issue 4:**

Collection floor icon on Android is broken

android (left) iOS (right)
image

**Issue 5:**

Cannot interact with video/gif on android in expanded view (media player options don't appear)

**Issue 6:**

Padding on smaller devices for collection floor price modal Got it button is off

(iPhone SE 2nd gen on left, iPhone 11 on right)
Screen Shot 2021-11-03 at 8 20 40 PM

**Comment by skylarbarrera** [comment]
i went ahead and fixed almost all current android issues, there are 2 extra's 
that we need to fix. 
https://github.com/rainbow-me/rainbow/pull/2519

File: src/components/expanded-state/unique-token/saveToCameraRoll.js (hunk 5 above)
Comment (skylarbarrera) [reply src/components/expanded-state/unique-token/saveToCameraRoll.js:7]: broken on android atm

File: src/components/expanded-state/unique-token/ZoomableWrapper.js (hunk 4 above)
Comment (osdnk) [reply src/components/expanded-state/unique-token/ZoomableWrapper.js:80]: any ideas? I have an opinion that this is actually good and the rules of hooks should be more lenient here.

**Comment by skylarbarrera** [comment]
just fixed 3d model & audio aspect ratios, they we're cut off with the existing logic.  added asset colors to audio view to match the buttons.

PoW 3d: https://cloud.skylarbarrera.com/Screen-Shot-2021-11-10-14-36-38.55.png
PoW audio: https://cloud.skylarbarrera.com/Screen-Shot-2021-11-10-14-48-34.79.png

this all looks good to me from a QA perspective, will do a code review tomorrow.

few things we should fix on android but non blocking:
- zoomable wrapper dont work and is swallowing all tappable events on android, video controls, modals, audio, etc
- blur effect doesnt work on android, we may want to just take the image color and do a light gradient instead

**Review by brunobar79 — APPROVED** [review]
Just fixed the conflicts and locked the dep versions. Good 2 go!

**Review by skylarbarrera — APPROVED** [review]
:shipit:

**Pull request merged**